### ISSUE TITLE
Add support for the maxDataPoints parameter

### ISSUE THREAD

**Issue by avereha** [dgryski/carbonapi#91]
It is documented here:

http://graphite.readthedocs.org/en/latest/render_api.html

and it seems to be used by grafana:

http://docs.grafana.org/installation/performance/

**Comment by avereha** [comment]
We probably also need the 'consolidateBy' function.

**Comment by szibis** [comment]
I think to achieve best performance this should be handled by carbonzipper/carbonserver and arguments passed from carbonapi->zipper ?
With this scenario we can get smaller number of data from carbonzipper which can get smaller number of data from carbonserver depends on maxdatapoints defined from carbonapi.

**Comment by nnuss** [comment]
`maxDataPoints` (as presently implemented in graphite) only has meaning for format=json and the effect is to `consolidate()` just before sending the response. The danger I could see is by `consolidate()`-ing before applying functions you potentially alter the result (particularly vs. graphite&#8482;).

**Comment by dgryski** [comment]
We now have support for consolidateBy().  Having maxDataPoints aggregate the metrics before rendering should be "easy".

**Comment by arodland** [comment]
"easy" version done here: https://github.com/arodland/carbonapi/tree/maxdatapoints

It's failing for me sometimes in production though, looks like `aggMean` is returning NaNs at unexpected times. I will investigate further when I have time.

**Comment by arodland** [comment]
I figured out the unexpected behavior, the linked branch now works for me. The problem wasn't the NaNs, the problem was that it's not valid to use `IsAbsent` together with `AggregatedValues`. Instead I took a page from cairo's book and trusted `IsNaN`/`IsInf`.

Is this approach okay? Should there be an `AggregatedIsAbsent` instead that reports on whether all of the points behind an aggregated point are absent?

If it is okay, should it be applied to the other formats besides JSON?

**Comment by dgryski** [comment]
Yes, post aggregation there isn't an equivalent for IsAbsent. I think we should fix that.

**Comment by nnuss** [comment]
I'm in favor of having maxDataPoints apply to all formats.

**Comment by arodland** [comment]
I added `AggregatedIsAbsent` and reworked things in terms of that. I tried to apply the same changes to the other serializers, but got a bit lost because I'm not familiar with them. I also didn't make any effort to unify things with Cairo, but there's probably some opportunity there. Finally, the tests fail because `marshalJSON` now expects access to a request (as cairo does). Other than that, it works well enough in real life with Grafana. If I make a PR of what I've got, can someone else take it to completion?
